File: internal/provider/kaggle/preflight.go
// Package kaggle contains a read-only preflight, not a dispatch-capable Provider.
// Account authentication is not evidence of private staging, GPU or batch readiness.
package kaggle

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"path/filepath"
	"reflect"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/vankhaivn/compute-relay/internal/credentials"
	"github.com/vankhaivn/compute-relay/internal/ports"
)

const (
	Local         Mode = "local"
	ReadOnly      Mode = "read_only"
	PythonSeries       = "3.11"
	ClientVersion      = "2.2.4"
	SDKVersion         = "0.1.35"
)

type Mode string

var (
	ErrConfig      = errors.New("invalid Kaggle preflight configuration")
	ErrMode        = errors.New("explicit local or read-only preflight mode is required")
	ErrProcess     = errors.New("Kaggle preflight process unavailable")
	ErrProtocol    = errors.New("invalid Kaggle preflight response")
	idPattern      = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$`)
	accountPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]{1,49}$`)
)

// Config is immutable, non-secret instance configuration. AccountName is a
// conservative canonical subset, not a claim about all upstream username rules.
// The interpreter must be the absolute path in the pinned client environment.
// No arbitrary command, endpoint, environment override or credential value exists.
type Config struct {
	InstanceID       string              `json:"instance_id"`
	Revision         string              `json:"revision"`
	AccountName      string              `json:"account_name"`
	CredentialRef    ports.CredentialRef `json:"credential_ref"`
	PythonExecutable string              `json:"python_executable"`
}

func (c Config) Validate() error {
	_, validRef := credentials.EnvironmentName(c.CredentialRef)
	if !idPattern.MatchString(c.InstanceID) || !idPattern.MatchString(c.Revision) ||
		!accountPattern.MatchString(c.AccountName) || !validRef ||
		!filepath.IsAbs(c.PythonExecutable) || strings.ContainsAny(c.PythonExecutable, "\x00\r\n\ufffd") {
		return ErrConfig
	}
	switch strings.ToLower(filepath.Ext(c.PythonExecutable)) {
	case ".cmd", ".bat", ".ps1", ".sh":
		return ErrConfig
	}
	return nil
}
func ParseConfig(raw []byte) (Config, error) {
	var c Config
	if len(raw) == 0 || len(raw) > 8192 || closedObject(raw, &c) != nil || c.Validate() != nil {
		return Config{}, ErrConfig
	}
	return c, nil
}

// Report contains only validated enums; no provider response, token, username,
// process output or exception text is returned. Local="ready" means exact pins
// matched. BatchReady must remain false until separately evidenced integration.
type Report struct {
	Protocol       int    `json:"protocol"`
	Mode           Mode   `json:"mode"`
	Local          string `json:"local"`
	Authentication string `json:"authentication"`
	AccountBinding string `json:"account_binding"`
	Quota          string `json:"quota"`
	BatchReady     bool   `json:"batch_ready"`
	Problem        string `json:"problem"`
}

func baseline(mode Mode) Report {
	return Report{Protocol: 1, Mode: mode, Local: "ready", Authentication: "not_checked", AccountBinding: "not_checked", Quota: "not_checked", Problem: "none"}
}
func (r Report) valid(mode Mode) bool {
	if r.Protocol != 1 || r.Mode != mode || (mode != Local && mode != ReadOnly) || r.BatchReady {
		return false
	}
	switch r.Local {
	case "unavailable", "version_mismatch":
		return r.Authentication == "not_checked" && r.AccountBinding == "not_checked" && r.Quota == "not_checked" && r.Problem == r.Local
	case "ready":
	default:
		return false
	}
	if mode == Local {
		return r == baseline(Local)
	}
	switch r.Problem {
	case "credential_unavailable", "credential_rejected":
		return r.Authentication == "failed" && r.AccountBinding == "not_checked" && r.Quota == "not_checked"
	case "access_denied", "provider_unavailable", "invalid_response":
		return r.Authentication == "unavailable" && r.AccountBinding == "not_checked" && r.Quota == "not_checked"
	case "account_mismatch":
		return r.Authentication == "verified" && r.AccountBinding == "mismatch" && r.Quota == "not_checked"
	case "quota_unavailable":
		return r.Authentication == "verified" && r.AccountBinding == "matched" && r.Quota == "unavailable"
	case "none":
		return r.Authentication == "verified" && r.AccountBinding == "matched" && (r.Quota == "available" || r.Quota == "unknown")
	default:
		return false
	}
}

// closedObject rejects duplicate keys before ordinary decoding can discard them.
// All fields are flat scalar values; unknown fields and trailing data are rejected.
func closedObject(raw []byte, target any) error {
	if !utf8.Valid(raw) {
		return ErrProtocol
	}
	allowed := map[string]bool{}
	typ := reflect.TypeOf(target).Elem()
	for i := 0; i < typ.NumField(); i++ {
		allowed[typ.Field(i).Tag.Get("json")] = true
	}
	d := json.NewDecoder(bytes.NewReader(raw))
	t, err := d.Token()
	if err != nil || t != json.Delim('{') {
		return ErrProtocol
	}
	seen := map[string]bool{}
	for d.More() {
		t, err = d.Token()
		key, ok := t.(string)
		if err != nil || !ok || seen[key] || !allowed[key] {
			return ErrProtocol
		}
		seen[key] = true
		var v json.RawMessage
		if d.Decode(&v) != nil || bytes.Equal(v, []byte("null")) || len(v) == 0 || v[0] == '{' || v[0] == '[' {
			return ErrProtocol
		}
	}
	if _, err = d.Token(); err != nil {
		return ErrProtocol
	}
	if d.Decode(new(any)) != io.EOF {
		return ErrProtocol
	}
	if len(seen) != len(allowed) {
		return ErrProtocol
	}
	d = json.NewDecoder(bytes.NewReader(raw))
	d.DisallowUnknownFields()
	if d.Decode(target) != nil {
		return ErrProtocol
	}
	return nil
}

type invocation func(context.Context, Config, Mode, []byte) (Report, error)
type Preflight struct {
	config   Config
	resolver ports.CredentialResolver
	run      invocation
	slot     chan struct{}
}

func New(c Config, resolver ports.CredentialResolver) (*Preflight, error) {
	if c.Validate() != nil || resolver == nil {
		return nil, ErrConfig
	}
	return &Preflight{config: c, resolver: resolver, run: runPython, slot: make(chan struct{}, 1)}, nil
}

// Check never runs automatically. ReadOnly is an explicit network/credential opt-in,
// not a dispatch permit. Each call checks local pins before looking up any secret.
// One invocation at a time per instance; a 60-second budget includes local checks,
// waiting for the slot, credential resolution and the two read-only SDK requests.
func (p *Preflight) Check(ctx context.Context, mode Mode) (Report, error) {
	if p == nil {
		return Report{}, ErrConfig
	}
	if mode != Local && mode != ReadOnly {
		return Report{}, ErrMode
	}
	ctx, cancel := context.WithTimeout(ctx, time.Minute)
	defer cancel()
	select {
	case p.slot <- struct{}{}:
		defer func() { <-p.slot }()
	case <-ctx.Done():
		return Report{}, ctx.Err()
	}
	if err := ctx.Err(); err != nil {
		return Report{}, err
	}
	r, err := p.run(ctx, p.config, Local, nil)
	if err != nil {
		return Report{}, err
	}
	if err := ctx.Err(); err != nil {
		return Report{}, err
	}
	if !r.valid(Local) {
		return Report{}, ErrProtocol
	}
	r.Mode = mode
	if mode == Local || r.Local != "ready" {
		return r, nil
	}
	invoked := false
	err = p.resolver.WithCredential(ctx, p.config.CredentialRef, func(secret []byte) error {
		if invoked {
			return ErrProtocol
		}
		if len(secret) == 0 || len(secret) > credentials.MaxBytes {
			return credentials.ErrUnavailable
		}
		for _, b := range secret {
			if b < 33 || b > 126 {
				return credentials.ErrUnavailable
			}
		}
		if err := ctx.Err(); err != nil {
			return err
		}
		invoked = true
		r, err = p.run(ctx, p.config, ReadOnly, secret)
		return err
	})
	if ctx.Err() != nil {
		return Report{}, ctx.Err()
	}
	if err != nil {
		// A resolver or child error may embed a secret; never wrap or reflect it.
		if invoked {
			return Report{}, ErrProcess
		}
		r = baseline(ReadOnly)
		r.Authentication = "failed"
		r.Problem = "credential_unavailable"
		return r, nil
	}
	if !invoked || !r.valid(ReadOnly) {
		return Report{}, ErrProtocol
	}
	return r, nil
}

File: internal/provider/kaggle/preflight.py
"""Fixed one-shot helper. No CLI parser, workload, browser or mutation entry point."""
import contextlib
import importlib.metadata
import json
import os
import re
import sys
import threading

PROTOCOL = 1
MAX_TOKEN = 8192
MAX_RESPONSE = 65536
ACCOUNT = re.compile(r"[a-z0-9][a-z0-9_-]{1,49}\Z")
URLS = (
    "https://api.kaggle.com/v1/security.OAuthService/IntrospectToken",
    "https://api.kaggle.com/v1/kernels.KernelsApiService/GetAcceleratorQuotaStatistics",
)


def baseline(mode):
    return dict(protocol=PROTOCOL, mode=mode, local="ready",
                authentication="not_checked", account_binding="not_checked",
                quota="not_checked", batch_ready=False, problem="none")


def local_check(mode):
    report = baseline(mode)
    try:
        exact = (sys.version_info[:2] == (3, 11)
                 and importlib.metadata.version("kaggle") == "2.2.4"
                 and importlib.metadata.version("kagglesdk") == "0.1.35")
        if not exact:
            report.update(local="version_mismatch", problem="version_mismatch")
    except Exception:
        report.update(local="unavailable", problem="unavailable")
    return report


def harden_session(client):
    # Version-bound SDK v0.1.35 seam: public http_client(), private session field.
    # Fail closed if the reviewed layout changes. Service calls and wire marshalling
    # stay in the official SDK; the hook only restricts its requests transport.
    import requests
    session = getattr(client.http_client(), "_session", None)
    if type(session) is not requests.Session:
        raise ValueError("unrecognized SDK session")
    session.trust_env = False
    session.max_redirects = 0
    session.cookies.clear()
    session.mount("https://", requests.adapters.HTTPAdapter(max_retries=0))
    count = 0

    def send(request, **kwargs):
        nonlocal count
        if count >= len(URLS) or request.method != "POST" or request.url != URLS[count]:
            raise ValueError("unexpected preflight request")
        count += 1
        # Explicit trust settings; never forward proxy/cert/redirect overrides.
        # Invoke the requests adapter directly: Session.send can pre-read a
        # redirect body even with allow_redirects=False. The adapter neither
        # follows redirects nor consumes a response before this byte guard.
        response = session.get_adapter(request.url).send(
            request, timeout=(5, 10), stream=True, proxies={}, verify=True, cert=None)
        try:
            if not 200 <= response.status_code < 300:
                response.raise_for_status()
                raise ValueError("redirect rejected")
            if response.headers.get("Content-Type", "").split(";", 1)[0].strip().lower() != "application/json":
                raise ValueError("unexpected content type")
            declared = response.headers.get("Content-Length")
            if declared is not None and (not declared.isdecimal() or int(declared) > MAX_RESPONSE):
                raise ValueError("oversized response")
            data = bytearray()
            for chunk in response.iter_content(chunk_size=16384):
                if len(chunk) > MAX_RESPONSE - len(data):
                    raise ValueError("oversized response")
                data.extend(chunk)
            def unique(items):
                result = {}
                for key, value in items:
                    if key in result:
                        raise ValueError("duplicate response field")
                    result[key] = value
                return result
            def invalid_constant(_):
                raise ValueError("invalid JSON constant")
            parsed = json.loads(data.decode("utf-8"), object_pairs_hook=unique,
                                parse_constant=invalid_constant)
            if not isinstance(parsed, dict):
                raise ValueError("expected response object")
            # Requests' own Response cache is populated only after bounded EOF.
            # The SDK reads it through its ordinary response parser after return.
            response._content = bytes(data)
            response._content_consumed = True
            return response
        finally:
            response.close()

    session.send = send


def read_only(token, account):
    report = baseline("read_only")
    from kagglesdk.kaggle_client import KaggleClient
    from kagglesdk.kaggle_env import KaggleEnv
    from kagglesdk.security.types.oauth_service import IntrospectTokenRequest
    from kagglesdk.kernels.types.kernels_api_service import ApiGetAcceleratorQuotaStatisticsRequest
    import requests
    try:
        # A nonempty explicit token prevents SDK environment/file auth fallback.
        # PROD is fixed; do not import kaggle, whose top-level import authenticates.
        with KaggleClient(env=KaggleEnv.PROD, verbose=False, api_token=token) as client:
            harden_session(client)
            request = IntrospectTokenRequest()
            request.token = token
            identity = client.security.oauth_client.introspect_token(request)
            if identity.active is not True:
                return dict(report, authentication="failed", problem="credential_rejected")
            username = identity.username
            if not isinstance(username, str) or not ACCOUNT.fullmatch(username):
                return dict(report, authentication="unavailable", problem="invalid_response")
            report["authentication"] = "verified"
            if username != account:
                return dict(report, account_binding="mismatch", problem="account_mismatch")
            report["account_binding"] = "matched"
            try:
                quota = client.kernels.kernels_api_client.get_accelerator_quota_statistics(
                    ApiGetAcceleratorQuotaStatisticsRequest())
                # Only endpoint/data availability here. Precision, units, age and
                # scheduler integration belong to M4-04, not an invented allowance.
                report["quota"] = "unknown" if quota.gpu_quota is None else "available"
            except Exception:
                report.update(quota="unavailable", problem="quota_unavailable")
            return report
    except requests.exceptions.HTTPError as exc:
        status = getattr(getattr(exc, "response", None), "status_code", None)
        if status == 401:
            return dict(report, authentication="failed", problem="credential_rejected")
        problem = "access_denied" if status == 403 else "provider_unavailable"
    except Exception:
        problem = "provider_unavailable"
    return dict(baseline("read_only"), authentication="unavailable", problem=problem)


def main():
    if len(sys.argv) != 3 or sys.argv[1] not in ("local", "read_only") or not ACCOUNT.fullmatch(sys.argv[2]):
        return 2
    mode, account = sys.argv[1:]
    report = local_check(mode)
    # Keep diagnostics out of the machine protocol; exception text can contain a
    # token, provider body or host path. The parent never reflects raw output either.
    with open(os.devnull, "w", encoding="utf-8") as quiet:
        with contextlib.redirect_stdout(quiet), contextlib.redirect_stderr(quiet):
            if mode == "read_only" and report["local"] == "ready":
                raw = sys.stdin.buffer.read(MAX_TOKEN + 1)
                if not raw or len(raw) > MAX_TOKEN or any(b < 33 or b > 126 for b in raw):
                    report.update(authentication="failed", problem="credential_unavailable")
                else:
                    try:
                        report = read_only(raw.decode("ascii"), account)
                    except Exception:
                        report.update(authentication="unavailable", problem="provider_unavailable")
    print(json.dumps(report, sort_keys=True, separators=(",", ":")))
    return 0


def bounded_main(seconds=25):
    # Independent wall limit: a dead Go parent must not leave the fixed helper
    # waiting indefinitely on stdin or a dribbling provider connection.
    timer = threading.Timer(seconds, lambda: os._exit(3))
    timer.daemon = True
    timer.start()
    try:
        return main()
    finally:
        timer.cancel()


if __name__ == "__main__":
    raise SystemExit(bounded_main())
